internal/datastore/crdb/tx.go

```diff
@@ -83,26 +91,68 @@ func execute(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transac
 		return nil
 	}
 
-	for i = 0; i < maxRetries; i++ {
+	for i = 0; i < maxRetries+1; i++ {
 		if err = releasedFn(tx); err != nil {
-			if !retriable(ctx, err) {
-				return err
+			if retriable(ctx, err) {
+				if _, retryErr := tx.Exec(ctx, "ROLLBACK TO SAVEPOINT cockroach_restart"); retryErr != nil {
```

Comment:
maybe move this logic into a function and then just do `rtx, rerr := retry(...)` and assign the `rtx` to `tx` if the retry succeeded? It would then allow the reset to be its own thing as well.